web: Pause player when tab is inactive (close #347)

OmarShehata opened the pull request:
This resolves https://github.com/ruffle-rs/ruffle/issues/347. It modifies `ruffle-player.ts` to detect when the tab is inactive using the browser's [visibility API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API). It also adds a `is_playing()` function to the public API in `web/src/lib.rs`. This is to allow the player to only resume playback if the animation wasn't already paused.

I considered whether there should be an option on the player to turn off this pausing, but since it is not currently possible to continue playing the frames in the background, and catching up may not be feasible, disabling this would produce incorrect behavior. 

It may make sense to revisit toggling this behavior when frame playback is controlled on a worker as mentioned here: https://github.com/ruffle-rs/ruffle/issues/347#issuecomment-763024001

> Possibly this can be done as we switch over to web workers, but I suspect this will be problematic on e.g. Safari which has no AudioWorklet.

### How I tested this

* Dragged and dropped an SWF into the `web/packages/demo` project using this branch
* Confirmed switching tab pauses it and audio stays in sync with animation
* Confirmed pausing the player directly will not unpause when switching tabs unless the user clicks the big play button.
* Tested in Chrome & Firefox on Windows 10

OmarShehata commented on web/src/lib.rs:
```diff
@@ -224,14 +223,24 @@ impl Ruffle {
 
     pub fn pause(&mut self) {
-        // Remove instance from the active list.
         INSTANCES.with(|instances| {
```
I removed these two comments because they seem to be a copy/paste error from the `destroy` function.

Herschel commented on web/packages/core/src/ruffle-player.ts:
```diff
@@ -165,9 +166,59 @@ export class RufflePlayer extends HTMLElement {
+        const documentAsAny: any = (document as any);
+        if (typeof documentAsAny.hidden !== "undefined") { // Opera 12.10 and Firefox 18 and later support
+          hidden = "hidden";
+          visibilityChange = "visibilitychange";
+        } else if (typeof documentAsAny.msHidden !== "undefined") {
+          hidden = "msHidden";
+          visibilityChange = "msvisibilitychange";
+        } else if (typeof documentAsAny.webkitHidden !== "undefined") {
+          hidden = "webkitHidden";
+          visibilityChange = "webkitvisibilitychange";
+        }
+
+        // Warn if the browser doesn't support addEventListener or the Page Visibility API
+        if (typeof document.addEventListener === "undefined" || hidden === undefined) {
+          console.log("The Page Visibility API is not supported in this browser. Audio may be out of sync when switching tabs.");
+        } else {
```
These support checks are overkill because if we're on a browser that doesn't support document.addEventListener, we'd fail much earlier than this point. That is, anything that doesn't support [Page Visibility](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event#browser_compatibility) already doesn't support [WebAssembly](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiate) or Shadow DOM or etc., so it seems fine to just expect the standard API, and not check for prefixes.

Herschel commented on web/src/lib.rs:
```diff
@@ -224,14 +223,24 @@ impl Ruffle {
         });
     }
 
+    pub fn is_playing(&mut self) -> JsValue {
```
You can return a standard rust `bool` here and `wasm_bindgen` takes care of converting it to JS:
```
        INSTANCES.with(|instances| {
            let instances = instances.borrow();
            let instance = instances.get(self.0).unwrap();
            let is_playing = instance.borrow().core.lock().unwrap().is_playing();
            is_playing
        })
```

Herschel reviewed (changes requested):
Thanks for this, functionally it seems great!

Could you fix the eslint errors as well? `npx eslint . --fix`

OmarShehata commented:
Thanks for the quick feedback @Herschel ! My bad for missing the eslint. I've fixed that, and addressed the other feedback comments.

I force pushed to keep the git history clean (it's not explicitly advised [in the contributing guide](https://github.com/ruffle-rs/ruffle/blob/master/CONTRIBUTING.md#commit-message-guidelines) but I thought it was implied). 

I re-tested with Chrome & Firefox on Windows 10 to make sure the behavior is still the same.

Herschel commented:
Thank you!
